modules/ui/fields/address.js

```diff
@@ -36,91 +36,94 @@ export function uiFieldAddress(field, context) {
             .filter(isAddressable)
-            .map(function(d) {
-                var loc = context.projection([
-                    (extent[0][0] + extent[1][0]) / 2,
-                    (extent[0][1] + extent[1][1]) / 2
-                ]);
-                var choice = geoChooseEdge(context.graph().childNodes(d), loc, context.projection);
-
+            .map(d => {
+                let dist = geoSphericalDistance(d.extent(context.graph()).center(), l);
+
+                if (d.type === 'way') {
+                    var loc = context.projection([
+                        (extent[0][0] + extent[1][0]) / 2,
+                        (extent[0][1] + extent[1][1]) / 2
+                    ]);
+                    var choice = geoChooseEdge(context.graph().childNodes(d), loc, context.projection);
+                    dist = Math.min(dist, choice.distance);
+                }
+
+                const value = resultProp && d.tags[resultProp] ? d.tags[resultProp] : d.tags.name;
+                let title = value;
+                if (type === 'street') {
+                    title = `${addrField.t('placeholders.street')}: ${title}`;
+                } else if (type === 'place') {
+                    title = `${addrField.t('placeholders.place')}: ${title}`;
+                }
                 return {
-                    title: d.tags.name,
-                    value: d.tags.name,
-                    dist: choice.distance
+                    title,
+                    value,
+                    dist,
+                    type,
+                    klass: `address-${type}`
                 };
             })
             .sort(function(a, b) {
                 return a.dist - b.dist;
             });
 
-        return utilArrayUniqBy(streets, 'value');
+        return utilArrayUniqBy(features, 'value');
+    }
 
+    function getNearStreets() {
         function isAddressable(d) {
             return d.tags.highway && d.tags.name && d.type === 'way';
         }
-    }
-
 
-    function getNearCities() {
-        var extent = combinedEntityExtent();
-        var l = extent.center();
-        var box = geoExtent(l).padByMeters(200);
-
-        var cities = context.history().intersects(box)
-            .filter(isAddressable)
-            .map(function(d) {
-                return {
-                    title: d.tags['addr:city'] || d.tags.name,
-                    value: d.tags['addr:city'] || d.tags.name,
-                    dist: geoSphericalDistance(d.extent(context.graph()).center(), l)
-                };
-            })
-            .sort(function(a, b) {
-                return a.dist - b.dist;
-            });
+        return getNear(isAddressable, 'street', 200);
+    }
 
-        return utilArrayUniqBy(cities, 'value');
+    function getNearPlaces() {
+        function isAddressable(d) {
+            return d.tags.name && d.tags.place ||
+                d.tags.boundary === 'administrative' && d.tags.admin_level >= 8;
```

Comment:
The order of precedence here is pretty confusing without parentheses.

Reply:
good catch! as written `(d.tags.name && d.tags.place) || (d.tags.boundary === 'administrative' && d.tags.admin_level >= 8)` the logic is even not working as intended! it should be `d.tags.name && (d.tags.place || (d.tags.boundary === 'administrative' && d.tags.admin_level >= 8))` :bow: